src/core.js

```diff
@@ -0,0 +1,572 @@
+  constructor(mapper){
+    super();
+    this.mapper = mapper;
+  }
+  resolved(x){
+    return new Resolved(this.mapper(x));
+  }
+  toString(){
+    return `map(${showf(this.mapper)})`;
+  }
+}
+
+export class BimapAction extends Action{
+  constructor(lmapper, rmapper){
+    super();
+    this.lmapper = lmapper;
+    this.rmapper = rmapper;
+  }
+  rejected(x){
+    return new Rejected(this.lmapper(x));
+  }
+  resolved(x){
+    return new Resolved(this.rmapper(x));
+  }
+  toString(){
+    return `bimap(${showf(this.lmapper)}, ${showf(this.rmapper)})`;
+  }
+}
+
+export class ChainAction extends Action{
+  constructor(mapper){
+    super();
+    this.mapper = mapper;
+  }
+  resolved(x){
+    return this.mapper(x);
+  }
+  toString(){
+    return `chain(${showf(this.mapper)})`;
+  }
+}
+
+export class MapRejAction extends Action{
+  constructor(mapper){
+    super();
+    this.mapper = mapper;
+  }
+  rejected(x){
+    return new Rejected(this.mapper(x));
+  }
+  toString(){
+    return `mapRej(${showf(this.mapper)})`;
+  }
+}
+
+export class RaceAction extends Action{
+  constructor(other){
+    super();
+    this.other = other;
+  }
+  run(early){
+    return new RaceActionState(early, this.other);
+  }
+  toString(){
+    return `race(${this.other.toString()})`;
+  }
+}
+
+export class RaceActionState extends RaceAction{
+  constructor(early, other){
+    super(other);
+    this.cancel = other._fork(
+      x => early(new Rejected(x)),
+      x => early(new Resolved(x))
+    );
+  }
+  rejected(x){
+    this.cancel();
+    return new Rejected(x);
+  }
+  resolved(x){
+    this.cancel();
+    return new Resolved(x);
+  }
+}
+
+export class BothAction extends Action{
+  constructor(other){
+    super();
+    this.other = other;
+  }
+  run(early){
+    return new BothActionState(early, this.other);
+  }
+  toString(){
+    return `both(${this.other.toString()})`;
+  }
+}
+
+export class BothActionState extends BothAction{
+  constructor(early, other){
+    super(new Eager(other));
+    this.cancel = this.other.cancel;
+  }
+  resolved(x){
+    return this.other.map(y => [x, y]);
+  }
+}
+
+export class OrAction extends Action{
+  constructor(other){
+    super();
+    this.other = other;
+  }
+  run(early){
+    return new OrActionState(early, this.other);
+  }
+  toString(){
+    return `or(${this.other.toString()})`;
+  }
+}
+
+export class OrActionState extends OrAction{
+  constructor(early, other){
+    super(new Eager(other));
+    this.cancel = this.other.cancel;
+  }
+  rejected(){
+    return this.other;
+  }
+  resolved(x){
+    this.cancel();
+    return new Resolved(x);
+  }
+}
+
+const Undetermined = 0;
+const Synchronous = 1;
+const Asynchronous = 2;
+
+export class Sequence extends Core{
+
+  constructor(spawn, actions = []){
+    super();
+    this._spawn = spawn;
+    this._actions = actions;
+  }
+
+  _transform(action){
+    return new Sequence(this._spawn, this._actions.concat([action]));
```

Comment:
Good point. This is the only place I think. This array doesn't see much action though, that's all handled by the double ended queue in _fork.